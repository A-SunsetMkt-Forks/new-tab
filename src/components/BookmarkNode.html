<!--
  FIXME: Try to prevent empty comment anchors from being added to the
  DOM since these bookmarks are static.
-->

{#if _node.children === undefined}

  <!-- bookmark link -->

  <LinkItem _url="{_node.url}" _title="{_node.title}" _max="{_max}" />
{:else}

  <!-- folder -->

  <div
    class="item folder {_endNode ? 'right' : ''}"
    title="{_node.title}"
    on:mouseenter="_handleMouseEnter()"
    on:mouseleave="_handleMouseLeave()"
  >
    {_node.title}

    {#if _lvl !== 0}
      <div class="caret">â–¸</div>
    {/if}

    {#if _show}

      <!-- subfolder -->

      <div class="subfolder {_endNode ? 'left' : ''}">
        {#each _node.children as _bookmark}
          <svelte:self _node="{_bookmark}" _lvl="{_lvl + 1}" _max="{40}" />
        {:else}
          <div class="item">(empty)</div>
        {/each}
      </div>
    {/if}
  </div>
{/if}


<script>
  import LinkItem from './LinkItem.html';

  const EVENT_NAME = 'open';
  const OPEN_DELAY = 200;
  const CLOSE_DELAY = 400;

  export default {
    components: {
      LinkItem,
    },
    // data: () => ({
    //   _show: false,
    //   _endNode: false,
    // }),
    oncreate() {
      this._openTimer = null;
      this._closeTimer = null;
    },
    ondestroy() {
      // clean up listener to prevent memory leaks
      if (this.listener) this.listener.cancel();
    },
    methods: {
      // set up event listener to close subfolder when another opens
      _onOpen() {
        this.listener = this.root.on(EVENT_NAME, (_lvl) => {
          if (this.get()._show) {
            if (_lvl === this.get()._lvl) {
              this.set({ _show: false });
              this._resetCloseTimer();
            }
          }
        });
      },

      // subfolder mouse enter
      _handleMouseEnter() {
        if (this.get()._show) {
          this._resetCloseTimer();
        } else {
          // delay to prevent accidental folder open
          this._openTimer = setTimeout(() => {
            this.root.fire(EVENT_NAME, this.get()._lvl);
            this.set({ _show: true });
            this._onOpen();
          }, OPEN_DELAY);
        }
      },

      // subfolder mouse leave
      _handleMouseLeave() {
        // reset open timer
        if (this._openTimer !== null) {
          clearTimeout(this._openTimer);
          this._openTimer = null;
        }

        // set up subfolder close timer
        if (this.get()._show) {
          this._closeTimer = setTimeout(() => {
            this.set({ _show: false });
          }, CLOSE_DELAY);
        }
      },

      _resetCloseTimer() {
        if (this._closeTimer !== null) {
          clearTimeout(this._closeTimer);
          this._closeTimer = null;
        }
      },
    },
  };
</script>

<style type="text/postcss" global>
  .folder {
    position: relative;
  }

  .caret {
    float: right;
    padding-left: 12px;
  }

  :global(.subfolder) {
    position: absolute;
    top: 40px;
    left: 0;
    background: var(--c0);
    box-shadow: var(--s0);

    & & {
      top: 0;
      left: 100%;
    }
  }

  /* bookmark item pushed to the right */
  .right {
    width: 100%;
    text-align: right;
  }

  /* folder which opens to the left */
  .left {
    right: 0;
    left: initial;
    text-align: initial;

    :global(.subfolder) {
      right: 100%;
      left: initial;
    }
  }
</style>
