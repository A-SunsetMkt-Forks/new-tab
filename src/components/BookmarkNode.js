import LinkItem from './LinkItem.html';
import BookmarkFolder from './BookmarkFolder.html';

export default function BookmarkNode(context) {
  /* eslint-disable-next-line no-underscore-dangle */
  return new (context.data._node.children !== undefined
    ? BookmarkFolder
    : LinkItem)(context);
}
