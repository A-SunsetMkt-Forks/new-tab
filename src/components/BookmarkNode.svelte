<script>
  import LinkItem from './LinkItem.svelte';
  import BookmarkFolder from './BookmarkFolder.svelte';
</script>

{#if $$props._node.children !== undefined}
  <BookmarkFolder {...$$props} />
{:else}
  <LinkItem {...$$props} />
{/if}
