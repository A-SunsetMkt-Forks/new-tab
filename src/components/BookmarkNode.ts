// eslint-disable-next-line import/no-extraneous-dependencies
import { SvelteComponent } from 'svelte/internal';
import LinkItem from './LinkItem.svelte';
import BookmarkFolder from './BookmarkFolder.svelte';

export default (context: ComponentOptions): SvelteComponent =>
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  new (context.props!.node.children !== undefined ? BookmarkFolder : LinkItem)(
    context,
  );
