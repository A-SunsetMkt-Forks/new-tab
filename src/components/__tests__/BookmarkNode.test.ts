import BookmarkNode from '../BookmarkNode';

const node = {};

describe('BookmarkNode component', () => {
  it('renders correctly', () => {
    expect.assertions(1);
    const target = document.createElement('div');
    // new BookmarkNode({
    BookmarkNode({
      props: { node },
      target,
    });
    expect(target.innerHTML).toMatchSnapshot();
  });
});
