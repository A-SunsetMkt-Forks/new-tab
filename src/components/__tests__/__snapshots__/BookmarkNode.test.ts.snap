// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookmarkNode component renders correctly 1`] = `
"<a><img src=\\"chrome://favicon/undefined\\">
  undefined</a>"
`;
